compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1441,6 +1441,15 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                 record!(self.tables.unused_generic_params[def_id.to_def_id()] <- unused);
             }
         }
+
+        // Encode all the deduced parameter attributes for everything that has MIR, even for items
+        // that can't be inlined.
+        for &local_def_id in tcx.mir_keys(()) {
+            if let DefKind::AssocFn | DefKind::Fn = tcx.def_kind(local_def_id) {
+                record_array!(self.tables.deduced_param_attrs[local_def_id.to_def_id()] <-
+                    self.tcx.deduced_param_attrs(local_def_id.to_def_id()));
+            }
+        }
```

Comment:
Encoding should be conditional on `tcx.sess.opts.output_types.should_codegen()` to avoid querying optimized MIR for check profile.

Reply:
I would have suggested putting this in the `encode_opt` branch above.

Reply:
The readonly attribute is also added at a call site, so it might be useful even if you don't encode optimized MIR of a particular function (non-generic, non-inline).

Reply:
Right, I originally was going to do that, but I saw that `encode_opt` only runs for inlineable functions, which isn't as aggressive as we could be.